### ISSUE TITLE
Does Velox support compiling and debugging on M1 chip Mac ?

### ISSUE THREAD

**Issue by frankliee** [facebookincubator/velox#1562]

**Comment by mbasmanova** [comment]
@Yuhta Jimmy is working on this.

**Comment by mbasmanova** [comment]
See PR #1528

**Comment by frankliee** [comment]
Thanks. I guess many people want to develop Velox on M1 Mac.

**Comment by kgpai** [comment]
@frankliee Yes this is supported now. Feel free to open this issue should you encounter any issues.

**Comment by usurai** [comment]
Hi, thanks for the effort for supporting the M1.
However, I still cannot get Velox up and running on M1 due to errors thrown by `setup-macos.sh`, there are two issues:
1. permission denied when installing range-v3
2. folly build failure due to 

I use M1 mac running macOS 12.4, my clang is `Apple clang version 13.1.6 (clang-1316.0.21.2.5)`.
When executing `CPU_TARGET="arm64" ./scripts/setup-macos.sh`, I first got a permission issue:
```
-- Build files have been written to: /Users/r/w/velox/range-v3/_build
+ ninja -C _build install
ninja: Entering directory `_build'
[0/1] Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/lib/cmake/range-v3/range-v3-targets.cmake
CMake Error at cmake_install.cmake:53 (file):
  file INSTALL cannot copy file
  "/Users/r/w/velox/range-v3/_build/CMakeFiles/Export/lib/cmake/range-v3/range-v3-targets.cmake"
  to "/usr/local/lib/cmake/range-v3/range-v3-targets.cmake": Permission
  denied.

FAILED: CMakeFiles/install.util 
cd /Users/r/w/velox/range-v3/_build && /opt/homebrew/Cellar/cmake/3.23.2/bin/cmake -P cmake_install.cmake
ninja: build stopped: subcommand failed.
```

This has also been mentioned in other threads, I added `sudo` to `ninja` command to overcome this:
```
In cmake_install of setup-helper-functions.sh

ninja -C "${BINARY_DIR}" install -> sudo ninja -C "${BINARY_DIR}" install
```

Then when building `folly`, I got `Undefined symbols for architecture arm64`:
```
[289/290] Linking CXX executable folly/logging/example/logging_example
FAILED: folly/logging/example/logging_example 
: && /Library/Developer/CommandLineTools/usr/bin/c++ -mavx2 -mfma -mavx -mf16c -masm=intel -mlzcnt -std=c++17 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names  folly/logging/example/CMakeFiles/logging_example.dir/main.cpp.o -o folly/logging/example/logging_example  -Wl,-rpath,/opt/homebrew/lib  folly/logging/example/liblogging_example_lib.a  libfolly.a  /usr/local/lib/libfmt.a  /opt/homebrew/lib/libboost_context-mt.dylib  /opt/homebrew/lib/libboost_filesystem-mt.dylib  /opt/homebrew/lib/libboost_atomic-mt.dylib  /opt/homebrew/lib/libboost_program_options-mt.dylib  /opt/homebrew/lib/libboost_regex-mt.dylib  /opt/homebrew/lib/libboost_system-mt.dylib  /opt/homebrew/lib/libboost_thread-mt.dylib  /usr/local/lib/libdouble-conversion.a  /opt/homebrew/lib/libgflags.2.2.2.dylib  /opt/homebrew/lib/libglog.dylib  /opt/homebrew/lib/libevent.dylib  /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/lib/libz.tbd  /opt/homebrew/opt/openssl@1.1/lib/libssl.dylib  /opt/homebrew/opt/openssl@1.1/lib/libcrypto.dylib  /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/lib/libbz2.tbd  /opt/homebrew/lib/liblz4.dylib  /opt/homebrew/lib/libzstd.dylib  /opt/homebrew/lib/libsnappy.dylib  -lc++abi && :
Undefined symbols for architecture arm64:
  "fmt::v8::detail::assert_fail(char const*, int, char const*)", referenced from:
      std::__1::make_unsigned<long>::type fmt::v8::detail::to_unsigned<long>(long) in libfolly.a(Singleton.cpp.o)
      std::__1::make_unsigned<int>::type fmt::v8::detail::to_unsigned<int>(int) in libfolly.a(Singleton.cpp.o)
      std::__1::make_unsigned<unsigned long>::type fmt::v8::detail::to_unsigned<unsigned long>(unsigned long) in libfolly.a(Singleton.cpp.o)
      std::__1::make_unsigned<long>::type fmt::v8::detail::to_unsigned<long>(long) in libfolly.a(File.cpp.o)
      std::__1::make_unsigned<int>::type fmt::v8::detail::to_unsigned<int>(int) in libfolly.a(File.cpp.o)
      fmt::v8::detail::format_decimal_result<char*> fmt::v8::detail::format_decimal<char, unsigned long long>(char*, unsigned long long, int) in libfolly.a(File.cpp.o)
      std::__1::make_unsigned<unsigned long>::type fmt::v8::detail::to_unsigned<unsigned long>(unsigned long) in libfolly.a(File.cpp.o)
      ...
  "fmt::v8::detail::error_handler::on_error(char const*)", referenced from:
      int fmt::v8::detail::get_dynamic_spec<fmt::v8::detail::width_checker, fmt::v8::basic_format_arg<fmt::v8::basic_format_context<std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char> >, fmt::v8::detail::error_handler>(fmt::v8::basic_format_arg<fmt::v8::basic_format_context<std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char> >, fmt::v8::detail::error_handler) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<int, 0>(int) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<unsigned int, 0>(unsigned int) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<long long, 0>(long long) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<unsigned long long, 0>(unsigned long long) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<__int128, 0>(__int128) in libfolly.a(Singleton.cpp.o)
      unsigned long long fmt::v8::detail::width_checker<fmt::v8::detail::error_handler>::operator()<unsigned __int128, 0>(unsigned __int128) in libfolly.a(Singleton.cpp.o)
      ...
  "fmt::v8::detail::thousands_sep_result<char> fmt::v8::detail::thousands_sep_impl<char>(fmt::v8::detail::locale_ref)", referenced from:
      fmt::v8::detail::thousands_sep_result<char> fmt::v8::detail::thousands_sep<char>(fmt::v8::detail::locale_ref) in libfolly.a(File.cpp.o)
  "fmt::v8::detail::throw_format_error(char const*)", referenced from:
      std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > fmt::v8::detail::write<char, std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >(std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char const*) in libfolly.a(File.cpp.o)
      std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > fmt::v8::detail::concat<fmt::v8::detail::spec_field<char, int, 1>, fmt::v8::detail::concat<fmt::v8::detail::text<char>, fmt::v8::detail::concat<fmt::v8::detail::spec_field<char, unsigned short, 2>, fmt::v8::detail::text<char> > > >::format<std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char const*, int, unsigned short>(std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char const* const&, int const&, unsigned short const&) const in libfolly.a(File.cpp.o)
      decltype(fp0.out()) fmt::v8::formatter<unsigned short, char, void>::format<fmt::v8::basic_format_context<std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char> >(unsigned short const&, fmt::v8::basic_format_context<std::__1::back_insert_iterator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, char>&) const in libfolly.a(File.cpp.o)
  "fmt::v8::vformat(fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<fmt::v8::appender, char> >)", referenced from:
      folly::LogStreamProcessor::vformatLogString(folly::Range<char const*>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<fmt::v8::appender, char> >, bool&)::'lambda'()::operator()() const in main.cpp.o
      folly::LogStreamProcessor::vformatLogString(folly::Range<char const*>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<fmt::v8::appender, char> >, bool&)::'lambda'()::operator()() const in liblogging_example_lib.a(lib.cpp.o)
      folly::CacheLocality::readFromSysfsTree(std::__1::function<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)> const&) in libfolly.a(CacheLocality.cpp.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

Thanks.

**Comment by kgpai** [comment]
Also FYI, I think the script to build folly isnt being invoked right, you shouldnt get '-mavx2 -mfma -mavx -mf16c -masm=intel -mlzcnt -std=c++17 -arch arm64 ' as that means compile for intel but for arm.

**Comment by usurai** [comment]
With the help of @kgpai, we found that the folly build failure was due to my change to `cmake_install` causing the invalidation of build flags. To do the setup on M1 Mac, change the owner of `/usr/local/lib` and `/usr/local/include` to you, and invoke `setup-macos.sh` with `CPU_TARGET="arm64"`.
Per @kgpai, there's been effort going on to move the installation out of '/usr/local/` to the build directory.

**Comment by mbasmanova** [comment]
@usurai Thank you for the update. I'm happy to hear you are unblocked. Thank you, @kgpai .

**Comment by honeyhexin** [comment]
Hi @kgpai , could you please give a more detailed explanation about how to resolve this issue? I still meet it when I build the latest Velox with M1. Thank you so much! 
BTW, I have tried change the permission and then run the scripts, but still meet this link error:
<img width="987" alt="image" src="https://github.com/user-attachments/assets/3c6c7641-6ea2-444d-8cd0-d401bb52831e">

**Comment by kgpai** [comment]
Adding to @majetideepak 's point looks like its coming from folly - are you using system folly (i.e preinstalled) or bundled (i.e velox building it as part of its build). You can try folly_SOURCE=BUNDLED to see if it helps. 
If you still face problems, open a new issue with detailed cmake logs . Thanks !
